torchaudio_contrib/functional.py

```diff
@@ -1,106 +1,133 @@
 
-def STFT(sig, n_fft=2048, hop=None, window=None, **kwargs):
-    """A wrapper for torch.stft with some preset parameters.
-    Returned value keeps both real and imageinary parts.
-    For STFT magnitude, see spectrogram
+    Args:
+        x (Tensor): Tensor of audio of size (channel, signal)
+            or (batch, channel, signal).
+        n_fft (int): FFT window size.
+        hop_length (int): Number audio of frames between STFT columns.
+        len_win (int): Size of stft window.
+        window (Tensor): 1-D tensor.
+        pad (int): Amount of padding to apply to signal.
+        pad_mode: padding method (see torch.nn.functional.pad).
+        **kwargs: Other torch.stft parameters, see torch.stft for more details.
 
-    input -> (batch, channel, time) or (channel, time)
-    output -> (batch, channel, freq, hop, complex) or (channel, freq, hop, complex)
+    Returns:
+        Tensor: (batch, channel, freq, hop, complex)
+            or (channel, freq, hop, complex)
 
     """
-    if hop is None:
-        hop = n_fft // 4
-    if window is None:
-        window = torch.hann_window(n_fft)
-
-    if sig.dim() == 3:
-        batch, channel, time = sig.size()
-        out_shape = [batch, channel, n_fft//2+1, -1, 2]
-    elif sig.dim() == 2:
-        channel, time = sig.size()
-        out_shape = [channel, n_fft//2+1, -1, 2]
+
+    # (!) Only 3D, 4D, 5D padding with non-constant
+    # padding are supported for now.
+    if pad > 0:
+        x = F.pad(x, (pad, pad), pad_mode)
+
+    if x.dim() == 3:
+        batch, channel, time = x.size()
+        out_shape = [batch, channel, n_fft // 2 + 1, -1, 2]
+    elif x.dim() == 2:
+        channel, time = x.size()
+        out_shape = [channel, n_fft // 2 + 1, -1, 2]
     else:
         raise ValueError('Input tensor dim() must be either 2 or 3.')
 
-    sig = sig.reshape(-1, time)
-    stft = torch.stft(sig, n_fft, hop, window=window, **kwargs)
-    stft = stft.reshape(out_shape)
+    x = x.reshape(-1, time)
+
+    stft_out = torch.stft(x, n_fft, hop_length, window=window,
+                          win_length=window.size(0), **kwargs)
+    stft_out = stft_out.reshape(out_shape)
+
+    return stft_out
+
 
-    return stft
+def stft(x, n_fft=2048, hop_length=None, len_win=None,
+         window=None, pad=0, pad_mode="reflect", **kwargs):
+    """
+    Wraps _stft setting default values and correct window device.
+    See torchaudio_contrib.STFT for more details.
+    """
+    n_fft, hop_length, window = stft_defaults(
+        n_fft, hop_length, len_win, window)
+    return _stft(x, n_fft=n_fft, hop_length=hop_length,
+                 window=window.to(x.device), pad=pad,
+                 pad_mode=pad_mode, **kwargs)
 
 
-def spectrogram(sig, n_fft=2048, hop=None, window=None, power=1.0, **kwargs):
+def complex_norm(x, power=1.0):
+    """
+    Normalize complex tensor.
     """
-    returns magnitude of spectrogram
+    return x.pow(2).sum(-1).pow(power / 2.0)
```

Comment:
```python
return torch.norm(x, power, dim=-1)
```
Ah, wait, the `power` is not meant for this, it's applying an exponent on top of the norm. It should be documented, and the most efficient path should be taken. I assume `torch.norm` can be faster than the manual thing, but if we want a different power than `1`, then the manual thing may be better. But if the power is `2`, we want the last step to be skipped (unless we trust that `.pow(1.)` is always a no-op, but maybe it can't, because it still has to copy the tensor since it's an out-of-place operation).

Reply:
Yeah, again was going off of [torchaudio](https://github.com/pytorch/audio/blob/ea7fcf080020f27da2f7675f01f94b84bae4c922/torchaudio/transforms.py#L213). Happy to use `torch.norm` making it clear that then power is `sum(abs(x)**power)**(1./power)`. Or if we want power to mean "power that the frobenius norm is taken to" then we can substitute `x.pow(2).sum(-1).pow(power / 2.0)` for `torch.norm(x, 2, -1).pow(power)` if `pow(1.)` is nice?

Reply:
> Happy to use `torch.norm` making it clear that then power is `sum(abs(x)**power)**(1./power)`

That's what I originally interpreted `power` to mean, but I think there's no use case for this :) The idea was to allow computing either a magnitude spectrogram (power=1) or power spectrogram (power=2) or compressed spectrogram (power<1) and merge this step with the complex-to-magnitude transformation for efficiency.

> we can substitute `x.pow(2).sum(-1).pow(power / 2.0)` for `torch.norm(x, 2, -1).pow(power)` if `pow(1.)` is nice?

No, because for `power=2`, we'd want to avoid taking the square root and squaring it again, and for `power not in (1,2)`, we'd want to integrate the powering into the square root.

Reply:
> That's what I originally interpreted power to mean, but I think there's no use case for this :) 
:+1: 

>No, because for power=2, we'd want to avoid taking the square root and squaring it again, and for power not in (1,2), we'd want to integrate the powering into the square root.

OK so then we stick with the manual approach? I don't know if `torch.norm` is faster though.

Reply:
Ideally we would time it and use whatever is the fastest for the three cases `power=1`, `power=2`, and everything else. I'd imagine that for `power=1`, `torch.norm()` may be the fastest, for `power=2`, the manual approach without the last `pow()`, and for everything else, the manual approach with `pow(2./power)` in the end.

Reply:
Ok so I've timed both approaches using:

```python
import torch
def time_op(op, device, loops=100):
    profile = False if device == 'cpu' else True
    with torch.autograd.profiler.profile(use_cuda=profile) as prof:
        for _ in range(loops):
            op(x)

    totals = prof.total_average()
    cpu = round(totals.cpu_time_total/loops, 1)
    cuda = round(totals.cuda_time_total/loops, 1)
    print('%s CPU time; %s CUDA time'%(cpu, cuda))

def compare_ops(op1, op2, device):
    time_op(op1, device)
    time_op(op2, device)


device = 'cuda'
x = torch.randn(1045, 3200).to(torch.device(device))

print('Power == 1')
op_11 = lambda z: torch.norm(z, 2, 1)
op_12 = lambda z: z.pow(2.).sum(1).pow(1./2. )

compare_ops(op_11, op_12, device)
```

And seems like using `torch.norm` is faster for gpu, but manual is faster on cpu.
```
# CUDA
Power == 1
8.9 CPU time; 45.8 CUDA time
22.5 CPU time; 120.0 CUDA time

Power == 2
13.8 CPU time; 50.4 CUDA time
14.1 CPU time; 117.0 CUDA time
```
and 

```
# CPU
Power == 1
2915.6 CPU time; 0.0 CUDA time
1313.0 CPU time; 0.0 CUDA time

Power == 2
2918.8 CPU time; 0.0 CUDA time
1296.8 CPU time; 0.0 CUDA time
```

Reply:
> On speed - surprisingly, it might be not the case.

@keunwoochoi I'm seeing manual is still faster with cpu, but not cuda. This is true for the downmixing too. Prioritize cuda in the hopes cpu `norm` is sped up? Have the function check device?

Reply:
+1 for prioritizing cuda (in general) (cc' @faroit)

Reply:
👍 yeah that's what I went with